bundles/org.openhab.binding.telegram/src/main/java/org/openhab/binding/telegram/internal/TelegramHandler.java

```diff
@@ -278,6 +289,99 @@ private int handleUpdates(List<Update> updates) {
                 }
 
+                // build and publish messageEvent trigger channel payload
+                JsonObject messageRaw = json.parse(gson.toJson(message)).getAsJsonObject();
+                JsonObject messagePayload = new JsonObject();
+                messagePayload.addProperty("message_id", message.messageId());
+                messagePayload.addProperty("from",
+                        String.join(" ", new String[] { message.from().firstName(), message.from().lastName() }));
+                messagePayload.addProperty("chat_id", message.chat().id());
+                if (messageRaw.has("text")) {
+                    messagePayload.addProperty("text", message.text());
+                }
+                if (messageRaw.has("animation")) {
+                    JsonObject animationPayload = messageRaw.getAsJsonObject("animation");
+                    String animationURL = getFullDownloadUrl(
+                            animationPayload.getAsJsonPrimitive("file_id").getAsString());
+                    animationPayload.addProperty("file_url", animationURL);
+                    messagePayload.addProperty("animation_url", animationURL);
+                    if (animationPayload.has("thumb")) {
+                        animationPayload.getAsJsonObject("thumb").addProperty("file_url", getFullDownloadUrl(
+                                animationPayload.getAsJsonObject("thumb").getAsJsonPrimitive("file_id").getAsString()));
+                    }
+                }
+                if (messageRaw.has("audio")) {
+                    JsonObject audioPayload = messageRaw.getAsJsonObject("audio");
+                    String audioURL = getFullDownloadUrl(audioPayload.getAsJsonPrimitive("file_id").getAsString());
+                    audioPayload.addProperty("file_url", audioURL);
+                    messagePayload.addProperty("audio_url", audioURL);
+                    if (audioPayload.has("thumb")) {
+                        audioPayload.getAsJsonObject("thumb").addProperty("file_url", getFullDownloadUrl(
+                                audioPayload.getAsJsonObject("thumb").getAsJsonPrimitive("file_id").getAsString()));
+                    }
+                }
+                if (messageRaw.has("document")) {
+                    JsonObject documentPayload = messageRaw.getAsJsonObject("document");
+                    String documentURL = getFullDownloadUrl(
+                            documentPayload.getAsJsonPrimitive("file_id").getAsString());
+                    documentPayload.addProperty("file_url", documentURL);
+                    messagePayload.addProperty("document_url", documentURL);
+                    if (documentPayload.has("thumb")) {
+                        documentPayload.getAsJsonObject("thumb").addProperty("file_url", getFullDownloadUrl(
+                                documentPayload.getAsJsonObject("thumb").getAsJsonPrimitive("file_id").getAsString()));
+                    }
```

Comment:
Similar code is used in several places. Please refactor this into a function and call that instead.

Reply:
Resolved in [7fe68eb](https://github.com/openhab/openhab-addons/pull/9251/commits/7fe68eb54b09bc6a88bc1fd434b335a63ac1ae82)